### DISCUSSION TITLE
Sorting Groups according to the list

### DISCUSSION THREAD

**Discussion by armindarvish** [armindarvish/consult-omni#1] [Ideas]
This is continuing discussion from consult-web's async branch on how to get better sorting of different groups. 

here is a link to discussion from consult-web: https://github.com/armindarvish/consult-web/discussions/29

**Comment by dasoju** [comment]
wondering if there is a solution 
iam using consult-omni with mdfind as source and the the results show the old files first
I was wondering if i can get the newest files first

**Reply by armindarvish** [reply]
@dasoju The ordering you are referring to is controlled by `mdfind` and not `consult-omni`. But, you can define a transform (or filter) function and reverse the order of results like this:

```elisp
(defun consult-omni--mdfind-transform (candidates &optional _query)
  "Reverse the order of CANDIDATES from `consult-omni-mdfind'."
  (nreverse candidates))
```

and then redefine the mdfind source with that function as transform/filter:

```elisp
(consult-omni-define-source "mdfind"
                            :narrow-char ?f
                            :category 'file
                            :type 'async
                            :require-match t
                            :face 'consult-omni-engine-title-face
                            :request #'consult-omni--mdfind-builder
                            :transform #'consult-omni--mdfind-transform
                            :on-preview #'consult-omni--mdfind-preview
                            :on-return #'identity
                            :on-callback #'consult-omni--mdfind-callback
                            :preview-key consult-omni-preview-key
                            :search-hist 'consult-omni--search-history
                            :select-hist 'consult-omni--selection-history
                            :group #'consult-omni--group-function
                            :enabled (lambda () (executable-find "mdfind"))
                            :sort nil
                            :interactive consult-omni-intereactive-commands-type
                            :annotate nil)

```

**Comment by Gleek** [comment]
Continuing the discussion from https://github.com/armindarvish/consult-omni/issues/42

>In short, the order of the items are determined by the order they arrive. Local sources tend to return results faster than web requests. In addition, web requests results often arrive in chunks, so sometimes the first few results from a web source arrive early but the remaining arrive later after other sources. Because of the way grouping works in a multi-source, these results are added to their own group, which makes it look like things are jumping around.

This is what I suspected. To experiment with this I added a timer before a `'dynamic` source would do a callback. My expectation was that since the results would be sent later for that source, the group would appear lower in the list. It instead, pushed it self to the top once the results arrived. Infact, it seemed to not matter when the results arrive for my new source, it always was on top.

Here's the code to replicate:


```elisp
(cl-defun consult-omni--test-results(input &rest args &key callback &allow-other-keys)

    (defvar consult-omni--test-timer nil)
    (when consult-omni--test-timer
      (cancel-timer consult-omni--test-timer))
    (let ((entries `(,(propertize "test 1" :source "Test")
                     ,(propertize "test 2" :source "Test"))))
      (setq consult-omni--test-timer (run-with-timer 2 nil callback  entries))
      nil))

(consult-omni-define-source "Testing"
                            :type 'dynamic
                            :require-match t
                            :face 'default
                            :request #'consult-omni--test-results
                            :on-return #'ignore
                            :on-preview #'ignore
                            :on-callback #'ignore
                            :enabled (lambda () t)
                            :group #'consult-omni--group-function)

(let ((consult-omni-multi-sources '("Apps" "Testing")))
  (consult-omni-multi))
```

Making it return first, by removing the timer and pairing it with a slower source, such as `fd`, also keeps it on the top position.

**Comment by armindarvish** [comment]
@Gleek That's a bit confusing. I need to find some time to dig into it, but when I tested it, I turned grouping off by "-- :g nil" and I could see that everything was ordered by the order they arrived.

Personally, I tend to use =vertico-next-group= in multi-source searches to quickly find the source I am interested in, rather than relying on the order of sources.

**Reply by Gleek** [reply]
Thanks for the tip about `vertico-next-group`! It's a small improvement, but I'm still aiming for a Spotlight-like experience where the top result is consistently the most relevant.

I'm planning to investigate @minad's suggestion about mutable insertions using end cons cells when I have time. Also, tagging @karthink as he mentioned working on this in the `consult-web` thread. He might have some insights. 

Appreciate you looking into this and creating `consult-omni` - it's already a very useful tool! I was inspired by your video and may need to pick your brain about it in the future!

**Reply by minad** [reply]
I've added async source support to `consult--multi`: 
https://github.com/minad/consult/commit/ada079d5932700a8819ace622ef4323e73983161

The sources and candidates retain their order if `:sort nil` is passed to `consult--multi`. If `:sort t` is passed, the candidates are sorted first by the UI and their order determines the order of the groups.
